modules/compiler/targets/host/source/HostPassMachinery.cpp

```diff
@@ -147,6 +166,180 @@ void HostPassMachinery::registerPassCallbacks() {
+  }
+
+  return false;
+}
+
+llvm::ModulePassManager HostPassMachinery::getLateTargetPasses() {
+  // We may have a situation where there were already opaque structs in the
+  // context associated with HostTarget which have the same name as those
+  // in the deserialized module. LLVM tries to resolve this name clash by
+  // introducing suffixes to the opaque structs in the deserialized module e.g.
+  // __mux_dma_event_t becomes __mux_dma_event_t.0. This is a problem since
+  // later passes may rely on the name __mux_dma_event_t to identify the type,
+  // so here we remap the structs.
+  llvm::ModulePassManager PM;
+  PM.addPass(compiler::utils::UniqueOpaqueStructsPass());
+  PM.addPass(compiler::utils::SimpleCallbackPass([](llvm::Module &m) {
+    // the llvm identifier is no longer needed by our host code
+    if (auto md = m.getNamedMetadata("llvm.ident")) {
+      md->dropAllReferences();
+      md->eraseFromParent();
+    }
+  }));
+#ifdef CA_ENABLE_DEBUG_SUPPORT
+  PM.addPass(compiler::utils::SimpleCallbackPass([&](llvm::Module &m) {
+    // Custom host options are set as module metadata as a testing aid
+    auto &ctx = m.getContext();
+    if (options.device_args.data()) {
+      if (auto md = m.getOrInsertNamedMetadata("host.build_options")) {
+        llvm::MDString *md_string =
+            llvm::MDString::get(ctx, options.device_args.data());
+        md->addOperand(llvm::MDNode::get(ctx, md_string));
+      }
+    }
+  }));
+#endif  // CA_ENABLE_DEBUG_SUPPORT
+  return PM;
+}
+
+llvm::ModulePassManager HostPassMachinery::getKernelFinalizationPasses(
+    cargo::array_view<compiler::BaseModule::SnapshotDetails> snapshots,
+    std::optional<std::string> unique_prefix) {
+  llvm::ModulePassManager PM;
+  compiler::BasePassPipelineTuner tuner(options);
+
+  // On host we have degenerate sub-groups i.e. sub-group == work-group.
+  tuner.degenerate_sub_groups = true;
+
+  // Forcibly compute the BuiltinInfoAnalysis so that cached retrievals work.
+  PM.addPass(llvm::RequireAnalysisPass<compiler::utils::BuiltinInfoAnalysis,
+                                       llvm::Module>());
+
+// Fix for alignment issues endemic on 32 bit ARM, but can also arise on 32 bit
+// X86. We want this pass to run early so it needs to process less instructions
+// and to avoid having to deal with the side effects of other passes.
+#if defined(UTILS_SYSTEM_32_BIT)
+  PM.addPass(compiler::utils::AlignModuleStructsPass());
+#endif
+
+  // Handle the generic address space
+  PM.addPass(llvm::createModuleToFunctionPassAdaptor(
+      compiler::utils::ReplaceAddressSpaceQualifierFunctionsPass()));
+
+  addPreVeczPasses(PM, tuner);
+
+  PM.addPass(vecz::RunVeczPass());
+
+  compiler::BaseModule::addSnapshotPassIfEnabled(
+      PM, "host", HOST_SNAPSHOT_VECTORIZED, snapshots);
+
+  addLateBuiltinsPasses(PM, tuner);
+
+  compiler::utils::HandleBarriersOptions HBOpts;
+  // Barriers with opt disabled are broken on 32 bit for some llvm versions, see
```

Comment:
I wonder if this is still true on 15+ - although I appreciate this is just moving code

Reply:
Good question - I might try dropping this and running CI over it. The ticket isn't super descriptive.